viewing_party/party.py

```diff
@@ -1,23 +1,184 @@
+
+    if not user_data["watched"]:
+        return 0.0
+    
+    total_rating = 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"] 
+            
+    average = total_rating / len(user_data["watched"])
+
+    return average
+
+def get_most_watched_genre(user_data):
+    genre_count = {"Fantasy": 0, "Intrigue": 0, "Action": 0} #change method so that other genres can be included
+
+    if not user_data["watched"]:
+        return None 
+    
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in genre_count:
+            genre_count[genre] += 1   
+    top_genre = max(genre_count, key=genre_count.get)
+
+    return top_genre
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+
+    movies_not_watched_by_friends = []
+
+    for movie_watched_by_user in user_data["watched"]:
+        watched_by_any_friend = False
+
+        # checks if the movie is has been watched by a friend
+        for friend in user_data["friends"]:
+            if movie_watched_by_user in friend["watched"]:
+                watched_by_any_friend = True
+                break
+
+        # If no friend has watched it, adds it to the final list
+        if not watched_by_any_friend:
+            movies_not_watched_by_friends.append(movie_watched_by_user)
+    
+    return movies_not_watched_by_friends
+
+def get_friends_unique_watched(user_data):
 
+    movies_watched_by_friends_not_user = []
+
+    # loops through each friend's watched list
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            # appends to list if 1) movie is not in the user's watched list and 2) not already in the result list
+            if movie not in user_data["watched"] and movie not in movies_watched_by_friends_not_user:
+                movies_watched_by_friends_not_user.append(movie)
+    
+    return movies_watched_by_friends_not_user
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
 
+def get_available_recs(user_data):
+    subscriptions = set(user_data.get("subscriptions", []))
+
+    recommended_movies = set()  # Using a set to ensure uniqueness
+    user_watched = set()  # Track movies the user has already watched
+
+    # Populate user_watched set with movie titles from user_data
+    for movie in user_data.get("watched"):
+        if "title" in movie:
+            user_watched.add(movie["title"])
+  
+    # Loop through each friend and their watched movies
+    for friend in user_data.get("friends"):
+        for movie in friend.get("watched"):
+                # Only add the movie if it's hosted on a subscribed platform and not watched by the user
+            if movie["host"] in subscriptions and movie["title"] not in user_watched:

```

Comment:
Looking back at the first requirements of this function, isn't there a function that completes part of them for us? Specifically these two: 

Determine a list of recommended movies. A movie should be added to this list if and only if:
- The user has not watched it
- At least one of the user's friends has watched
- **...**

This could clear up a lot of the code that we have in this function.